packages/common/src/proto-utils.ts

```diff
@@ -0,0 +1,103 @@
+import proto from '@temporalio/proto';
+import { fromProto3JSON, toProto3JSON } from 'proto3-json-serializer';
+import { patchProtobufRoot } from '@temporalio/proto/lib/patch-protobuf-root';
+
+export type History = proto.temporal.api.history.v1.IHistory;
+export type Payload = proto.temporal.api.common.v1.IPayload;
+
+// Cast to any because the generated proto module types are missing the lookupType method
+const patched = patchProtobufRoot(proto) as any;
+const historyType = patched.lookupType('temporal.api.history.v1.History');
+const payloadType = patched.lookupType('temporal.api.common.v1.Payload');
+
+function pascalCaseToConstantCase(s: string) {
+  return s.replace(/[^\b][A-Z]/g, (m) => `${m[0]}_${m[1]}`).toUpperCase();
+}
+
+function fixEnumValue<O extends Record<string, any>>(obj: O, attr: keyof O, prefix: string) {
+  return (
+    obj[attr] && {
+      [attr]: `${prefix}_${pascalCaseToConstantCase(obj[attr])}`,
+    }
+  );
+}
+
+function fixHistoryEvent(e: Record<string, any>) {
+  const type = Object.keys(e).find((k) => k.endsWith('EventAttributes'));
+  if (!type) {
+    throw new TypeError(`Missing attributes in history event: ${JSON.stringify(e)}`);
+  }
+
+  return {
+    ...e,
+    eventType: `EVENT_TYPE_${pascalCaseToConstantCase(e.eventType)}`,
+    [type]: {
+      ...e[type],
+      ...(e[type].taskQueue && {
+        taskQueue: { ...e[type].taskQueue, ...fixEnumValue(e[type].taskQueue, 'kind', 'TASK_QUEUE_KIND') },
+      }),
+      ...fixEnumValue(e[type], 'parentClosePolicy', 'PARENT_CLOSE_POLICY'),
+      ...fixEnumValue(e[type], 'workflowIdReusePolicy', 'WORKFLOW_ID_REUSE_POLICY'),
+      ...fixEnumValue(e[type], 'initiator', 'CONTINUE_AS_NEW_INITIATOR'),
+      ...fixEnumValue(e[type], 'retryState', 'RETRY_STATE'),
+      ...(e[type].childWorkflowExecutionFailureInfo && {
+        childWorkflowExecutionFailureInfo: {
+          ...e[type].childWorkflowExecutionFailureInfo,
+          ...fixEnumValue(e[type].childWorkflowExecutionFailureInfo, 'retryState', 'RETRY_STATE'),
+        },
+      }),
+    },
+  };
+}
+
+function fixHistory(h: Record<string, any>) {
+  return {
+    events: h.events.map(fixHistoryEvent),
+  };
+}
+
+/**
+ * Convert a proto JSON representation of History to a valid History object
+ */
+export function historyFromJSON(history: unknown): History {
```

Comment:
Would `history: { events: []object }` make sense here?

Reply:
No because typically users would feed the response of `JSON.parse` here

Reply:
Then the user can choose to cast and/or ignore types